src/buffer/out/textBuffer.cpp

```diff
@@ -1001,37 +954,55 @@ void TextBuffer::Reset()
 
     try
     {
-        BufferAllocator allocator{ newSize };
-
-        const auto currentSize = GetSize().Dimensions();
-        const auto attributes = GetCurrentAttributes();
-
         til::CoordType TopRow = 0; // new top row of the screen buffer
         if (newSize.height <= GetCursor().GetPosition().y)
         {
             TopRow = GetCursor().GetPosition().y - newSize.height + 1;
         }
-        const auto TopRowIndex = (GetFirstRowIndex() + TopRow) % currentSize.height;
+        const auto TopRowIndex = gsl::narrow_cast<size_t>(_firstRow + TopRow) % _storage.size();
 
-        // rotate rows until the top row is at index 0
-        std::rotate(_storage.begin(), _storage.begin() + TopRowIndex, _storage.end());
-        _SetFirstRowIndex(0);
-
-        // realloc in the Y direction
-        // remove rows if we're shrinking
-        _storage.resize(allocator.height());
+        std::vector<ROW> newStorage;
+        auto newBuffer = _allocateBuffer(newSize, _currentAttributes, newStorage);
 
-        // realloc in the X direction
-        for (auto& it : _storage)
+        // This basically imitates a std::rotate_copy(first, mid, last), but uses ROW::CopyRangeFrom() to do the copying.
         {
-            it.Resize(allocator.chars(), allocator.indices(), allocator.width(), attributes);
-            ++allocator;
+            const auto first = _storage.begin();
+            const auto last = _storage.end();
+            const auto mid = first + TopRowIndex;
+            auto dest = newStorage.begin();
+
+            std::span<ROW> sourceRanges[]{
+                { mid, last },
+                { first, mid },
+            };
+
+            // Ensure we don't copy more from `_storage` than fit into `newStorage`.
+            if (sourceRanges[0].size() > newStorage.size())
+            {
+                sourceRanges[0] = sourceRanges[0].subspan(0, newStorage.size());
+            }
+            if (const auto remaining = newStorage.size() - sourceRanges[0].size(); sourceRanges[1].size() > remaining)
+            {
+                sourceRanges[1] = sourceRanges[1].subspan(0, remaining);
+            }
+
+            for (const auto& sourceRange : sourceRanges)
+            {
+                for (const auto& oldRow : sourceRange)
+                {
+                    til::CoordType begin = 0;
+                    dest->CopyRangeFrom(0, til::CoordTypeMax, oldRow, begin, til::CoordTypeMax);

```

Comment:
begin never changes; sup with it?

Reply:
In the current version of `CopyRangeFrom` it returns the `columnEnd` of the `dest` row, while the `begin` parameter (also called `otherBegin`) is modified to be the `columnEnd` of the source row. This allows you to implement reflow, by calling `CopyRangeFrom` until either of the two integers has exceeded their corresponding row width.

The API design is terrible and could use `RowWriteState`-like struct... But I think that's a rather simple change to make and can be done at any point in time.